<!DOCTYPE html>
<html lang="en" itemscope itemtype="http://schema.org/WebPage">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>R packages and example statistical analysis I - STAT 818: Introduction to R</title>
  

<meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=yes"/>

<meta name="MobileOptimized" content="width"/>
<meta name="HandheldFriendly" content="true"/>


<meta name="applicable-device" content="pc,mobile">

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">

<meta name="mobile-web-app-capable" content="yes">

<meta name="author" content="yli8" />
  <meta name="description" content="" />

  <meta name="keywords" content="STAT 818, R" />






<meta name="generator" content="Hugo 0.111.3" />


<link rel="canonical" href="https://lyqglyqg.github.io/STAT-818/post/wk8/" />





<link rel="icon" href="/STAT-818/favicon.ico" />











<link rel="stylesheet" href="/STAT-818/sass/jane.min.de22abc00de44766eebd1054fd9e0b254b071f89d5019044f893c484a9249a8d.css" integrity="sha256-3iKrwA3kR2buvRBU/Z4LJUsHH4nVAZBE&#43;JPEhKkkmo0=" media="screen" crossorigin="anonymous">







<meta property="og:title" content="R packages and example statistical analysis I" />
<meta property="og:description" content="" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lyqglyqg.github.io/STAT-818/post/wk8/" /><meta property="article:section" content="post" />
<meta property="article:published_time" content="2023-03-22T00:00:00+00:00" />
<meta property="article:modified_time" content="2023-03-22T00:00:00+00:00" />
<meta itemprop="name" content="R packages and example statistical analysis I">
<meta itemprop="description" content=""><meta itemprop="datePublished" content="2023-03-22T00:00:00+00:00" />
<meta itemprop="dateModified" content="2023-03-22T00:00:00+00:00" />
<meta itemprop="wordCount" content="693">
<meta itemprop="keywords" content="edgeR,R packages,Gene differential analysis," /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="R packages and example statistical analysis I"/>
<meta name="twitter:description" content=""/>

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->




</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">STAT 818: Introduction to R</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/">Home</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/post/">Archives</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/categories/">Categories</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/tags/">Tags</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/about/">Couse Syllabus</a>
          
        
      </li><li class="mobile-menu-item">
        
          
          <div class="mobile-menu-parent">
            <span class="mobile-submenu-open"></span>
            <a href="https://lyqglyqg.github.io/STAT-818/">
              docs
            </a>
          </div>
          <ul class="mobile-submenu-list">
            
              <li>
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk7/">Data visualization</a>
              </li>
            
              <li>
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk6/">Functions, control statements, packaging and debugging</a>
              </li>
            
              <li>
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk4/">Import, Export and Manage Data</a>
              </li>
            
              <li>
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk2/">R Objects I: Vector, Matrix, and Array</a>
              </li>
            
              <li>
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk3/">R Objects II: List, Factor, and Data Frame</a>
              </li>
            
              <li>
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk8/">R packages and example statistical analysis I</a>
              </li>
            
              <li>
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk5/">Reproducible Research: Using R Markdown</a>
              </li>
            
          </ul>
        
      </li>
    

    
  </ul>
</nav>


  
    






  <link rel="stylesheet" href="/lib/photoswipe/photoswipe.min.css" />
  <link rel="stylesheet" href="/lib/photoswipe/default-skin/default-skin.min.css" />




<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

<div class="pswp__bg"></div>

<div class="pswp__scroll-wrap">
    
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    
    <div class="pswp__ui pswp__ui--hidden">
    <div class="pswp__top-bar">
      
      <div class="pswp__counter"></div>
      <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
      <button class="pswp__button pswp__button--share" title="Share"></button>
      <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
      <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>
      
      
      <div class="pswp__preloader">
        <div class="pswp__preloader__icn">
          <div class="pswp__preloader__cut">
            <div class="pswp__preloader__donut"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
      <div class="pswp__share-tooltip"></div>
    </div>
    <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
    </button>
    <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
    </button>
    <div class="pswp__caption">
      <div class="pswp__caption__center"></div>
    </div>
    </div>
    </div>
</div>

  

  

  

  <header id="header" class="header container">
    <div class="logo-wrapper">
  <a href="/" class="logo">
    
      STAT 818: Introduction to R
    
  </a>
</div>

<nav class="site-navbar">
  <ul id="menu" class="menu">
    
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/">Home</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/post/">Archives</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/categories/">Categories</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/tags/">Tags</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          
            <a class="menu-item-link" href="https://lyqglyqg.github.io/STAT-818/about/">Couse Syllabus</a>
          

        

      </li>
    
        <li class="menu-item">
        
          
          <a class="menu-item-link menu-parent" href="https://lyqglyqg.github.io/STAT-818/">docs</a>
          <ul class="submenu">
            
              <li class="submenu-item">
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk7/">Data visualization</a>
              </li>
            
              <li class="submenu-item">
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk6/">Functions, control statements, packaging and debugging</a>
              </li>
            
              <li class="submenu-item">
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk4/">Import, Export and Manage Data</a>
              </li>
            
              <li class="submenu-item">
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk2/">R Objects I: Vector, Matrix, and Array</a>
              </li>
            
              <li class="submenu-item">
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk3/">R Objects II: List, Factor, and Data Frame</a>
              </li>
            
              <li class="submenu-item">
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk8/">R packages and example statistical analysis I</a>
              </li>
            
              <li class="submenu-item">
                <a href="https://lyqglyqg.github.io/STAT-818/post/wk5/">Reproducible Research: Using R Markdown</a>
              </li>
            
          </ul>

        

      </li>
    

    
    

    
  </ul>
</nav>

  </header>

  <div id="mobile-panel">
    <main id="main" class="main bg-llight wallpaper">
      <div class="content-wrapper">
        <div id="content" class="content container">
          <article class="post bg-white">
    
    <header class="post-header">
      <h1 class="post-title">R packages and example statistical analysis I</h1>
      

      <div class="post-meta">
  <div class="post-meta-author">
    by
      yli8
    
  </div>

  <div class="post-meta-time">
    <time datetime="2023-03-22">
      2023-03-22
    </time>
  </div>

  


  <div class="post-meta__right">
    

    <div class="post-meta-category">
        <a href="https://lyqglyqg.github.io/STAT-818/categories/edger/"> edgeR </a>
          <a href="https://lyqglyqg.github.io/STAT-818/categories/r-packages/"> R packages </a>
          <a href="https://lyqglyqg.github.io/STAT-818/categories/gene-differential-analysis/"> Gene differential analysis </a>
          
      </div>


    
    


    
    
  </div>
</div>

    </header>

    
    
<div class="post-toc" id="post-toc">
  <h2 class="post-toc-title">Table of Contents</h2>
  <div class="post-toc-content">
    <nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a></li>
    <li><a href="#installation">Installation</a>
      <ul>
        <li><a href="#edger-vignette"><code>edgeR</code> vignette</a></li>
        <li><a href="#analysis-step-1-read-in-data">Analysis step 1: read in data</a></li>
        <li><a href="#creat-a-cancer-indicator-and-a-dgelist-object">Creat a cancer indicator and a &lsquo;DGEList&rsquo; object</a></li>
        <li><a href="#filtering-and-normalization">Filtering and normalization</a></li>
        <li><a href="#de-analysis">DE analysis</a></li>
        <li><a href="#calculate-adjusted-p-values-and-sort-the-output-table">Calculate adjusted p-values and sort the output table</a></li>
        <li><a href="#generate-plots-to-visualize-outputs">Generate plots to visualize outputs</a></li>
      </ul>
    </li>
  </ul>
</nav>
  </div>
</div>


    
    <div class="post-content">
      <p>We use an <em>R</em> package named <code>edgeR</code> to learn how to do gene differential expression (DE) analysis using RNAseq count data.</p>
<h2 id="introduction">Introduction</h2>
<p><code>edgeR</code> is a widely used R package for conducting DE analysis with RNAseq data. Its statistical analysis is based on negative binomial distribution. It provides Bayes estimation, generalized linear modelling, Fisher exact test, and Quasi-likelihood test.</p>
<h2 id="installation">Installation</h2>
<p>First, you need to know your <em>R</em> version. To check that, use</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">R.Version</span><span class="p">()</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="Rversion.png">
<p>If your <em>R</em> version is &gt;3.5, then install <code>edgeR</code> use</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">requireNamespace</span><span class="p">(</span><span class="s">&#34;BiocManager&#34;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;BiocManager&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&#34;edgeR&#34;</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>If your <em>R</em> version is &lt;3.5, then install <code>edgeR</code> use</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">source</span><span class="p">(</span><span class="s">&#34;http://bioconductor.org/biocLite.R&#34;</span>）
</span></span><span class="line"><span class="cl"><span class="nf">biocLite</span><span class="p">(</span><span class="s">&#39;edgeR&#39;</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="edger-vignette"><code>edgeR</code> vignette</h3>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="o">&gt;</span> <span class="nf">library</span><span class="p">(</span><span class="n">edgeR</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="o">&gt;</span> <span class="nf">edgeRUsersGuide</span><span class="p">()</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="edgeRv.png" width="60%" height="60%">
<h3 id="analysis-step-1-read-in-data">Analysis step 1: read in data</h3>
<p>The &ldquo;geneCounts.csv&rdquo; file contains RNAseq count data from 419 ovarian cancer patients from TCGA data and 88 normal controls from GTEX data.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">X_c</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="s">&#34;geneCounts.csv&#34;</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">X_c[1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="n">]</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="X_screenshot.png"> 
<h3 id="creat-a-cancer-indicator-and-a-dgelist-object">Creat a cancer indicator and a &lsquo;DGEList&rsquo; object</h3>
<p>First, creat a group indicator of cancer vs normal control, with 1 indicating ovarian cancer. If a rowname starts by &ldquo;TCGA&rdquo;, then it is RNAseq data from a cancer subject. If it starts by &ldquo;GTEX&rdquo;, it is a normal control.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">sample_vec</span> <span class="o">&lt;-</span> <span class="nf">sapply</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">X_c</span><span class="p">),</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">unlist</span><span class="p">(</span><span class="nf">strsplit</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s">&#34;-&#34;</span><span class="p">))</span><span class="n">[[1]][1]</span><span class="p">)</span> 
</span></span><span class="line"><span class="cl"><span class="n">group</span> <span class="o">&lt;-</span> <span class="nf">as.numeric</span><span class="p">(</span><span class="n">sample_vec</span> <span class="o">==</span> <span class="s">&#34;TCGA&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">table</span><span class="p">(</span><span class="n">group</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="grp.png"> 
<p><code>edgeR</code> can create a &lsquo;DGEList&rsquo; object that contains a table of counts (rows=features, columns=samples), group indicator, library size (optional) for the downstream DE analysis.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">y</span> <span class="o">&lt;-</span> <span class="nf">DGEList</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="nf">t</span><span class="p">(</span><span class="n">X_c</span><span class="p">),</span><span class="n">group</span><span class="o">=</span><span class="n">group</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="filtering-and-normalization">Filtering and normalization</h3>
<p><code>edgeR</code> can filter out low quality genes (genes with low counts) and can normalize the gene counts to scale the raw library sizes.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">keep</span> <span class="o">&lt;-</span> <span class="nf">filterByExpr</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>  
</span></span><span class="line"><span class="cl"><span class="n">y</span> <span class="o">&lt;-</span> <span class="n">y[keep</span><span class="p">,,</span><span class="n">keep.lib.sizes</span><span class="o">=</span><span class="kc">FALSE</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">y</span> <span class="o">&lt;-</span> <span class="nf">calcNormFactors</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">design</span> <span class="o">&lt;-</span> <span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="n">group</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">y</span> <span class="o">&lt;-</span> <span class="nf">estimateDisp</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">design</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="de-analysis">DE analysis</h3>
<p>Use the <code>glmQLFit</code> and <code>glmQLFTest</code> to run the DE analysis and get a p-value for each gene.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">fit</span> <span class="o">&lt;-</span> <span class="nf">glmQLFit</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="n">design</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">qlf</span> <span class="o">&lt;-</span> <span class="nf">glmQLFTest</span><span class="p">(</span><span class="n">fit</span><span class="p">,</span><span class="n">coef</span><span class="o">=</span><span class="m">2</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="calculate-adjusted-p-values-and-sort-the-output-table">Calculate adjusted p-values and sort the output table</h3>
<p>We need to control for multiple testing here (this is a statistical concept not required for this course), the following syntax calculate the false discovery rates (fdr or adjusted p-values) and sort the output table from the most significant to least significant genes.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">qlfT</span> <span class="o">&lt;-</span> <span class="n">qlf</span><span class="o">$</span><span class="n">table</span>
</span></span><span class="line"><span class="cl"><span class="n">qlfT_sort</span> <span class="o">&lt;-</span> <span class="n">qlfT</span><span class="nf">[order</span><span class="p">(</span><span class="n">qlfT[</span><span class="p">,</span><span class="m">4</span><span class="n">]</span><span class="p">),</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">qlfT_sort</span><span class="o">$</span><span class="n">fdr</span> <span class="o">&lt;-</span> <span class="nf">p.adjust</span><span class="p">(</span><span class="n">qlfT_sort[</span><span class="p">,</span><span class="m">4</span><span class="n">]</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span><span class="s">&#34;fdr&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">qlfT_sort[1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="n">]</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="outputTable.png"> 
<p>The smaller the <code>PValue</code> or the <code>fdr</code>, the more differently expressed for the gene between the ovarian cancer subjects and the normal controls.</p>
<p>The <code>logFC</code> is for &ldquo;log fold change&rdquo; with base 2. &ldquo;logFC=3.14&rdquo; means the average count number of the ovarian cancer group is $2^{3.14}$ times of that if the normal control group for the gene.</p>
<h3 id="generate-plots-to-visualize-outputs">Generate plots to visualize outputs</h3>
<p>First, we generate a heatmap for the most 200 differentially expressed genes (similar to what you did in the midterm exam).</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">topGene</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">qlfT_sort[1</span><span class="o">:</span><span class="m">200</span><span class="p">,</span><span class="n">]</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">idtmp</span> <span class="o">&lt;-</span> <span class="nf">match</span><span class="p">(</span><span class="n">topGene</span><span class="p">,</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">X_c</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">heatmapGene</span> <span class="o">&lt;-</span> <span class="n">X_c[</span> <span class="p">,</span> <span class="n">idtmp]</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">sample_names_short</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&#34;GTEX-&#34;</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">88</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">),</span> <span class="nf">paste</span><span class="p">(</span><span class="s">&#34;TCGA-&#34;</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">419</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#34;&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">group_col</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#34;GTEX&#34;</span><span class="p">,</span> <span class="m">88</span><span class="p">),</span> <span class="nf">rep</span><span class="p">(</span><span class="s">&#34;TCGA&#34;</span><span class="p">,</span> <span class="m">419</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">my_sample_col</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">sample</span> <span class="o">=</span> <span class="n">group_col</span><span class="p">)</span> <span class="c1">#sample_names_short)</span>
</span></span><span class="line"><span class="cl"><span class="nf">row.names</span><span class="p">(</span><span class="n">my_sample_col</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">sample_names_short</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="nf">row.names</span><span class="p">(</span><span class="n">heatmapGene</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">sample_names_short</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="nf">pheatmap</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">heatmapGene</span><span class="p">),</span> <span class="n">annotation_col</span> <span class="o">=</span> <span class="n">my_sample_col</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">cluster_rows</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span> <span class="n">cluster_cols</span> <span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="heatmap818.png"> 
<p>Next, we generate a plot called &ldquo;volvano plot&rdquo;, which can show both the scales of &ldquo;logFC&rdquo; and the significance of &ldquo;PValues&rdquo; (in term of $-log_10(PValue)$).</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">qlfT_sort_vol</span> <span class="o">&lt;-</span> <span class="n">qlfT_sort</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">  <span class="nf">EnhancedVolcano</span><span class="p">(</span><span class="n">qlfT_sort_vol</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">lab</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">qlfT_sort_vol</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="n">x</span> <span class="o">=</span> <span class="s">&#39;logFC&#39;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">y</span> <span class="o">=</span> <span class="s">&#39;PValue&#39;</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="volcano.png"> 
<p>One can specify more details to make the plots more informative. For example:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">keyvals</span> <span class="o">&lt;-</span> <span class="nf">rep</span><span class="p">(</span><span class="s">&#39;gray&#39;</span><span class="p">,</span> <span class="nf">dim</span><span class="p">(</span><span class="n">qlfT_sort_vol</span><span class="p">)</span><span class="n">[1]</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">keyvals[101</span><span class="o">:</span><span class="m">500</span><span class="n">]</span> <span class="o">&lt;-</span> <span class="s">&#39;royalblue&#39;</span>
</span></span><span class="line"><span class="cl"><span class="n">keyvals[1</span><span class="o">:</span><span class="m">100</span><span class="n">]</span> <span class="o">&lt;-</span>  <span class="s">&#39;gold&#39;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="nf">names</span><span class="p">(</span><span class="n">keyvals</span><span class="p">)</span><span class="n">[keyvals</span> <span class="o">==</span> <span class="s">&#39;gold&#39;</span><span class="n">]</span> <span class="o">&lt;-</span> <span class="s">&#39;Strong gene&#39;</span>
</span></span><span class="line"><span class="cl"><span class="nf">names</span><span class="p">(</span><span class="n">keyvals</span><span class="p">)</span><span class="n">[keyvals</span> <span class="o">==</span> <span class="s">&#39;royalblue&#39;</span><span class="n">]</span> <span class="o">&lt;-</span> <span class="s">&#39;Weak gene&#39;</span>
</span></span><span class="line"><span class="cl"><span class="nf">names</span><span class="p">(</span><span class="n">keyvals</span><span class="p">)</span><span class="n">[keyvals</span> <span class="o">==</span> <span class="s">&#39;gray&#39;</span><span class="n">]</span> <span class="o">&lt;-</span> <span class="s">&#39;No significant gene&#39;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">shreshold</span> <span class="o">&lt;-</span> <span class="m">6.14e-9</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">  <span class="nf">EnhancedVolcano</span><span class="p">(</span><span class="n">qlfT_sort_vol</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">lab</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">qlfT_sort_vol</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="n">x</span> <span class="o">=</span> <span class="s">&#39;logFC&#39;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">y</span> <span class="o">=</span> <span class="s">&#39;PValue&#39;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">selectLab</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">qlfT_sort_vol</span><span class="p">)</span><span class="nf">[which</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">keyvals</span><span class="p">)</span> <span class="o">%in%</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;Strong gene&#39;</span><span class="p">,</span> <span class="s">&#39;Weak gene&#39;</span><span class="p">))</span><span class="n">]</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">xlab</span> <span class="o">=</span> <span class="nf">bquote</span><span class="p">(</span><span class="o">~</span><span class="n">Log[2]</span><span class="o">~</span> <span class="s">&#39;fold change&#39;</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="n">title</span> <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">pCutoff</span> <span class="o">=</span> <span class="m">1.0e-55</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">FCcutoff</span> <span class="o">=</span> <span class="m">2.0</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">pointSize</span> <span class="o">=</span> <span class="m">3.0</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">labSize</span> <span class="o">=</span> <span class="m">3.0</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">colCustom</span> <span class="o">=</span> <span class="n">keyvals</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">hline</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">shreshold</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="n">hlineCol</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;red&#39;</span><span class="p">))</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="volcano_1.png">
    </div>

    
    



    
    


    <footer class="post-footer">
      <div class="post-tags">
          <a href="https://lyqglyqg.github.io/STAT-818/tags/edger/">edgeR</a>
            <a href="https://lyqglyqg.github.io/STAT-818/tags/r-packages/">R packages</a>
            <a href="https://lyqglyqg.github.io/STAT-818/tags/gene-differential-analysis/">Gene differential analysis</a>
            
        </div>


      
      <nav class="post-nav">
        
        
          <a class="next" href="/STAT-818/post/wk7/">
            <span class="next-text nav-default">Data visualization</span>
            <span class="prev-text nav-mobile">Next</span>
            
            <i class="iconfont">
              <svg class="icon" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="18" height="18">
  <path d="M332.091514 74.487481l-75.369571 89.491197c-10.963703 12.998035-10.285251 32.864502 1.499144 44.378743l286.278095 300.375162L266.565125 819.058374c-11.338233 12.190647-11.035334 32.285311 0.638543 44.850487l80.46666 86.564541c11.680017 12.583596 30.356378 12.893658 41.662889 0.716314l377.434212-421.426145c11.332093-12.183484 11.041474-32.266891-0.657986-44.844348l-80.46666-86.564541c-1.772366-1.910513-3.706415-3.533476-5.750981-4.877077L373.270379 71.774697C361.493148 60.273758 343.054193 61.470003 332.091514 74.487481z"></path>
</svg>

            </i>
          </a>
      </nav>
    </footer>
  </article>

  
  


  
  

  

  
  

  
  

  

  

    

  

  

        </div>
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="icon-links">
  
  
    <a href="mailto:yli8@kumc.edu" rel="me noopener" class="iconfont"
      title="email" >
      <svg class="icon" viewBox="0 0 1451 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="36" height="36">
  <path d="M664.781909 681.472759 0 97.881301C0 3.997201 71.046997 0 71.046997 0L474.477909 0 961.649408 0 1361.641813 0C1361.641813 0 1432.688811 3.997201 1432.688811 97.881301L771.345323 681.472759C771.345323 681.472759 764.482731 685.154773 753.594283 688.65053L753.594283 688.664858C741.602731 693.493018 729.424896 695.068979 718.077952 694.839748 706.731093 695.068979 694.553173 693.493018 682.561621 688.664858L682.561621 688.65053C671.644501 685.140446 664.781909 681.472759 664.781909 681.472759L664.781909 681.472759ZM718.063616 811.603883C693.779541 811.016482 658.879232 802.205449 619.10784 767.734955 542.989056 701.759633 0 212.052267 0 212.052267L0 942.809523C0 942.809523 0 1024 83.726336 1024L682.532949 1024 753.579947 1024 1348.948139 1024C1432.688811 1024 1432.688811 942.809523 1432.688811 942.809523L1432.688811 212.052267C1432.688811 212.052267 893.138176 701.759633 817.019477 767.734955 777.248 802.205449 742.347691 811.03081 718.063616 811.603883L718.063616 811.603883Z"></path>
</svg>

    </a>


<a href="https://lyqglyqg.github.io/STAT-818/index.xml" rel="noopener alternate" type="application/rss&#43;xml"
    class="iconfont" title="rss" target="_blank">
    <svg class="icon" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="30" height="30">
  <path d="M819.157333 1024C819.157333 574.592 449.408 204.8 0 204.8V0c561.706667 0 1024 462.293333 1024 1024h-204.842667zM140.416 743.04a140.8 140.8 0 0 1 140.501333 140.586667A140.928 140.928 0 0 1 140.074667 1024C62.72 1024 0 961.109333 0 883.626667s62.933333-140.544 140.416-140.586667zM678.784 1024h-199.04c0-263.210667-216.533333-479.786667-479.744-479.786667V345.173333c372.352 0 678.784 306.517333 678.784 678.826667z"></path>
</svg>

  </a>
  
</div>

<div class="copyright">
  <span class="power-by">
    Powered by <a class="hexo-link" href="https://gohugo.io">Hugo</a>
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    Theme - <a class="theme-link" href="https://github.com/xianmin/hugo-theme-jane">Jane</a>
  </span>

  <span class="copyright-year">
    &copy;
    2023
    <span class="heart">
      
      <i class="iconfont">
        <svg class="icon" viewBox="0 0 1025 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="14" height="14">
  <path d="M1000.1 247.9c-15.5-37.3-37.6-70.6-65.7-98.9-54.4-54.8-125.8-85-201-85-85.7 0-166 39-221.4 107.4C456.6 103 376.3 64 290.6 64c-75.1 0-146.5 30.4-201.1 85.6-28.2 28.5-50.4 61.9-65.8 99.3-16 38.8-24 79.9-23.6 122.2 0.7 91.7 40.1 177.2 108.1 234.8 3.1 2.6 6 5.1 8.9 7.8 14.9 13.4 58 52.8 112.6 102.7 93.5 85.5 209.9 191.9 257.5 234.2 7 6.1 15.8 9.5 24.9 9.5 9.2 0 18.1-3.4 24.9-9.5 34.5-30.7 105.8-95.9 181.4-165 74.2-67.8 150.9-138 195.8-178.2 69.5-57.9 109.6-144.4 109.9-237.3 0.1-42.5-8-83.6-24-122.2z"
   fill="#8a8a8a"></path>
</svg>

      </i>
    </span><span class="author">
        yli8
        
      </span></span>

  
  

  
</div>

    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont">
        
        <svg class="icon" viewBox="0 0 1024 1024" version="1.1"
  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
  width="35" height="35">
  <path d="M510.866688 227.694839 95.449397 629.218702l235.761562 0-2.057869 328.796468 362.40389 0L691.55698 628.188232l241.942331-3.089361L510.866688 227.694839zM63.840492 63.962777l894.052392 0 0 131.813095L63.840492 195.775872 63.840492 63.962777 63.840492 63.962777zM63.840492 63.962777"></path>
</svg>

      </i>
    </div>
  </div>
  
<script type="text/javascript" src="/STAT-818/lib/jquery/jquery-3.2.1.min.js"></script>
  <script type="text/javascript" src="/STAT-818/lib/slideout/slideout-1.0.1.min.js"></script>




<script type="text/javascript" src="/STAT-818/js/main.fe83e11b4fbc9193d67e2c9db78bad21f8dc59fca0cacd8c1c3bb071bb16a852.js" integrity="sha256-/oPhG0&#43;8kZPWfiydt4utIfjcWfygys2MHDuwcbsWqFI=" crossorigin="anonymous"></script>



  <script type="text/javascript">
    window.MathJax = {
      showProcessingMessages: false,
      messageStyle: 'none'
    };
  </script>
  <script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-MML-AM_CHTML' async></script>









  
    <script type="text/javascript" src="/js/load-photoswipe.js"></script>
    <script type="text/javascript" src="/lib/photoswipe/photoswipe.min.js"></script>
    <script type="text/javascript" src="/lib/photoswipe/photoswipe-ui-default.min.js"></script>
  

















</body>
</html>
